src/libs/migrations/CheckForPreviousReportActionID.js

```diff
@@ -0,0 +1,61 @@
+import _ from 'underscore';
+import Onyx from 'react-native-onyx';
+import Log from '../Log';
+import ONYXKEYS from '../../ONYXKEYS';
+
+/**
+ * @returns {Promise<Object>}
+ */
+function getReportActionsFromOnyx() {
+    return new Promise((resolve) => {
+        const connectionID = Onyx.connect({
+            key: ONYXKEYS.COLLECTION.REPORT_ACTIONS,
+            waitForCollectionCallback: true,
+            callback: (allReportActions) => {
+                Onyx.disconnect(connectionID);
+                return resolve(allReportActions);
+            },
+        });
+    });
+}
+
+/**
+ * Migrate Onyx data for reportActions. If the first reportAction of a reportActionsForReport
+ * does not contain a 'previousReportActionID', all reportActions for all reports are removed from Onyx.
+ *
+ * @returns {Promise<void>}
+ */
+export default function () {
+    return getReportActionsFromOnyx().then((allReportActions) => {
+        if (_.isEmpty(allReportActions)) {
+            Log.info(`[Migrate Onyx] Skipped migration CheckForPreviousReportActionID because there were no reportActions`);
+            return;
+        }
+
+        const firstValidReportAction = _.find(_.values(allReportActions), (reportActions) => !_.isEmpty(reportActions));
```

Comment:
`!_.isEmpty(reportActions)` doesn't mean it will have a valid report action @neonbhai 

What if the input is like this, this case it will fail to identify?
you could add a inner _.find here or basically run `for` loop on `_.values(allReportActions)` and find the `firstValidValue`

```
r1: {
     1: {
              message: 1,
       },
       2: {
               message: 2,
       },
},
r2: {
    1: {
        reportActionID: 1,
     }
}

```

Reply:
Was assuming all `reportActions` would have a `reportActionID` key and if we have some keys, that would indicate the reportAction loaded correctly?

If not,

Since we cannot get the exact `reportActionData` that passes the inner check, as the outer find returns the first passing `reportActions` object. To get the first valid value we would have to execute a find again. Would this be satisfactory?

```javascript
const firstValidReportAction = _.find(_.values(allReportActions), (reportActions) => _.find(_.values(reportActions), (reportActionData) => _.has(reportActionData, 'reportActionID')));

const firstValidValue = _.find(_.values(firstValidReportAction), (reportActionData) => _.has(reportActionData, 'reportActionID'));
```

Reply:
>To get the first valid value we would have to execute a find again. Would this be satisfactory?
 
 Or we can basically run for loop on _.values(allReportActions) as suggested?

>do you think it would be safe to assume if we have some keys, that would mean reportAction integrity is maintained

We are making sure we don't miss any edge cases in case of zoombie reportActions. Since reportActionID is a mandatory key it should better to check if its present or not.

Reply:
@neonbhai ^

Reply:
@abdulrahuman5196 Added these. Please check the changes :bow: